hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/service/KeyDeletingService.java

```diff
@@ -338,24 +398,137 @@ private void processSnapshotDeepClean(int delCount)
               if (delCount < keyLimitPerTask) {
                 // Deep clean is completed, we can update the SnapInfo.
                 deepCleanedSnapshots.add(currSnapInfo.getTableKey());
+                // exclusiveSizeList contains check is used to prevent
+                // case where there is no entry in deletedTable, this
+                // will throw NPE when we submit request.
+                if (previousSnapshot != null && exclusiveSizeMap
+                    .containsKey(previousSnapshot.getTableKey())) {
+                  completedExclusiveSizeSet.add(
+                      previousSnapshot.getTableKey());
+                }
               }
 
               if (!keysToPurge.isEmpty()) {
                 processKeyDeletes(keysToPurge, currOmSnapshot.getKeyManager(),
                     keysToModify, currSnapInfo.getTableKey());
               }
             } finally {
-              if (previousSnapshot != null) {
-                rcPrevOmSnapshot.close();
-              }
+              IOUtils.closeQuietly(rcPrevOmSnapshot, rcPrevToPrevOmSnapshot);
             }
           }
 
         }
       }
+
+      updateSnapshotExclusiveSize();
       updateDeepCleanedSnapshots(deepCleanedSnapshots);
     }
 
+    @SuppressWarnings("checkstyle:ParameterNumber")
+    private void calculateExclusiveSize(
+        SnapshotInfo previousSnapshot,
+        SnapshotInfo previousToPrevSnapshot,
+        OmKeyInfo keyInfo,
+        OmBucketInfo bucketInfo, long volumeId,
+        Table<String, String> snapRenamedTable,
+        Table<String, OmKeyInfo> previousKeyTable,
+        Table<String, String> prevRenamedTable,
+        Table<String, OmKeyInfo> previousToPrevKeyTable) throws IOException {
+      String prevSnapKey = previousSnapshot.getTableKey();
+      long exclusiveReplicatedSize =
+          exclusiveReplicatedSizeMap.getOrDefault(
+              prevSnapKey, 0L) + keyInfo.getReplicatedSize();
+      long exclusiveSize = exclusiveSizeMap.getOrDefault(
+          prevSnapKey, 0L) + keyInfo.getDataSize();
```

Comment:
Default is supposed to be 0 or current exclusive size of snapshot? For example, key1 is share between snapshot1 and snapshot2. When one of the snapshot gets deleted, key1 will be become exclusive for other snapshot. In that case, it should be on top of existing exclusive size for the snapshot.

Reply:
Ah, Got it. :+1: